### ISSUE TITLE
Android support

### ISSUE THREAD

**Issue by jlnr** [jlnr/gosu#273]
Sorry, this is just a brainstorming ticket since I am not sure where to go with Android.

I have ported bits of C++ Gosu to Android during one Ludum Dare competition:

https://github.com/jlnr/ld26

It worked, but I had no clue about Android development at the time. And parts of the Gosu "port" live in the ld26 repository, while some other parts live in Gosu's Android branch.
Plus that was Gosu 0.7.x, not Gosu 0.8.x, which is based on SDL 2.x, which in turn has an Android port.

If I want to get this to work again, it will probably need a full rewrite, and it would be wise to delegate as much work to SDL 2 as possible.

There is also this project, which looks really nice: https://github.com/Garoe/gosu-android
Example game here: https://github.com/rubendahl/cult_of_gems

And RubyMotion works on Android now, so that's another thing I'd love to play with. Whatever I do in the C++ world, it would be cool if it was possible to use it from Ruby as well (either Robuto or RubyMotion).

**Comment by ashes999** [comment]
+1

**Comment by ashes999** [comment]
I learned about this through the original author of `gosu-android`. Their solution is basically to wrap up Java-specific calls with Ruby, because Ruboto uses JRuby. Sadly, `gosu-android` is no longer maintained.

I am very, very interested in a free, non-Java-based Ruby solution for Android. MRuby looks solid (and has SDL 2 and OpenGL support), although I lack the C/C++ skills to use it.

Please keep me posted. Ruboto is awesome, although I don't know how well it works with gems that have native (C++) components, like Gosu does. I know they requested a new `dalvik` gem platform specifically for JRuby-compatible extensions. 

FYI, this is for my [Terrace project](http://github.com/ashes999/terrace/), which is an attempt at a cross-platform Ruby game engine (write one flavour of Ruby code, and generate web, desktop, and mobile versions).

**Comment by jlnr** [comment]
Gosu on Android is up next on my list (almost - the iOS port needs a bit of love, and the CMake files require a couple of hotfixes). I have plans for a few cross-platform mobile games, without the web port. If I could only free up a weekend or two for coding...

So far I've been looking at RubyMotion for Android, as I'm already using it on iOS, but wrapping Gosu in Java first and then using it from Ruby sounds painful. Maybe I'll give mruby another try, thanks for bringing it up again. (I don't care if the Ruby implementation is free or not, I'll just go for the lesser evil in terms of programming :) )

**Comment by ashes999** [comment]
RubyMotion only works on Mac, so that eliminates me (and probably a lot of developers).

Ruboto is not as bad as it sounds. It's JRuby, so it's not really wrapped code, but Ruby in Java. It has its own strengths, like real threading. I [think I almost have libGDX working](https://github.com/ruboto/ruboto/issues/743) with JRuby on Android (it was pretty easy to get it working with JRuby on desktop).

**Comment by jlnr** [comment]
I have some experience with JRuby on the desktop & servers, but I think I'd rather give mruby a try (next to RubyMotion). More C and less Java :)

**Comment by andrewhavens** [comment]
Not sure if you heard, but it was recently announced that the next version of RubyMotion will come with a cross-platform game library called motion-game. I've heard its based on cocos2d. Anyway, might be something to keep an eye on in terms of Gosu-RubyMotion integration. https://www.rubygamedev.com/posts/36

**Comment by elieteyssedou** [comment]
Hi ! I was present to the last conference of RubyMotion (#inspect2015) in Paris, the last week. I could talk with the CTO of Ruby Motion, and there is nothing official but yes motion-game is possibly coming ! Actually the game engine is under development by the founder of RubyMotion himself, but at the beginning he did it only for him.
The thing is that he took a few minutes to present it at #inspect2015, so I think that it could be serious. Actually it was VERY VERY cool ! The point, --> if I remember well <-- , is that motion-game is planned for iOs for the moment.

**Comment by jlnr** [comment]
Thanks for the heads-up, I'll check it out. I guess it won't be compatible with Gosu at all, as most libraries seem to deal with node-based scene graphs (retained mode) instead of complete redraws (immediate mode). :( I'd love not having to port my C++ backend to Android and then wrapping it, but I don't see a good way around it for now.

**Comment by jlnr** [comment]
Just a random thought, but maybe wrapping Gosu for use with Java (for RubyMotion) would also allow it to be used with Robuto(?) and, more importantly, JRuby on Windows. Three platforms with one wrapper! So much research to do :)

**Comment by nightblade9** [comment]
Any updates on this? It's been three years since the last comment (one year if you count the linked issue).

**Comment by jlnr** [comment]
No progress.

If anyone can figure out how to make Ruby work on other platforms (and not just a proof of concept, but a long-lived ecosystem), I'll happily port Gosu. But RubyMotion on Android doesn't look very healthy to me, and mruby doesn't seem to have entered the mainstream yet either. I'm kind of hoping that [DragonRuby](https://dragonruby.itch.io) will take off, and that Gosu can piggy-back on its toolchain at some point. (Not talking about Android specifically! There's not a single platform where distributing Ruby games is easy.)

**Comment by shawn42** [comment]
The bummer is that DragonRuby is closed source and my guess is they will not release anything that really helps Gosu.  :(

**Comment by nightblade9** [comment]
I agree with @shawn42. DragonRuby is closed source. Based on other posts by the author, I think they're using `mruby` (he has a post about an embedded `mruby` interpreter on Switch).

Ruby2D is not bad. They also use `mruby`; my understanding of their stack is: Simple2D (C/C++ engine with OpenGL/SDL?), which they create wrappers for in Ruby, and expose through `mruby`. 

`mruby` hasn't really entered mainstream since its release, and most of the early projects/gems they list on their page are defunct (years gone by without updates). What the community really needs, is a single good, ongoing project (like `gosu`) to figure this out and push it forward.

I have a tiny bit of background in `mruby`, a bit more in Ruby, and a fair amount in Android. I would be happy to help with this. C/C++, OpenGL, and SDL are not my strong suites, so I can't figure it out on my own.

@jlnr I think, ultimately, it's up to you to decide if you want to do this. If so, I will help however I can. The best thing would be if someone exposed an intermediate-level library (just `mruby` on, say, SDL or OpenGL, and an easy way to build projects for web, PC, and mobile) which gosu et all could leverage. But it seems like most people hide that in their C/C++ layer, and then just expose Ruby bindings on top.

**Comment by jlnr** [comment]
Thanks for offering your help! I'll be time-constrained for the rest of the year, so it would be more realistic for me to support someone else working on this.

It's also hard for me to be optimistic about an Android port, given that there has been zero interest in Gosu for iOS, even now that RubyMotion is free. (Of course it's still closed source, reference-counted, and all that.) I think an mruby-based toolchain would have to be _extremely_ polished to take off, and it might be easier to tell people to port their games to Crystal when they want to publish them, but Crystal's Windows port is forever a work in progress.

The only way I can see myself investing time into Ruby/Crystal tooling again is if I can find the time to work on a couch-coop game, then port it to all Steam platforms with either mruby or Crystal, and then maybe the Android port will "fall into place" thanks to SDL.

**Comment by Kjarrigan** [comment]
> it might be easier to tell people to port their games to Crystal when they want to publish them (...)
> investing time into Ruby/Crystal tooling again

@jlnr Since you brought this up I'm a bit curious - have you "ported" or done anything with Gosu and Crystal? Every couple weeks I do a little bit with Crystal and it looks really promising. The have some built-in `FFI` thing. Would you go for that or building a "native" thing and transcribing gosu in Crystal.

**Comment by jlnr** [comment]
@Kjarrigan Not really. I don't want to invest into the language until the Windows port is completed. I keep looking at their docs, though. Maybe I'll start learning Crystal by writing an API backend for a client at some point.
